packages/core/core-v2/src/AssetGraph.js

```diff
@@ -0,0 +1,204 @@
+
+// @flow
+'use strict';
+import Graph, { Node, type NodeId } from './Graph';
+import type { Dependency, Asset, File } from './types';
+
+export const nodeFromRootDir = (rootDir: string) => ({
+  id: rootDir,
+  type: 'root',
+  value: rootDir,
+});
+
+export const nodeFromDep = (dep: Dependency) => ({
+  id: `${dep.sourcePath}:${dep.moduleSpecifier}`,
+  type: 'dependency',
+  value: dep,
+});
+
+export const nodeFromFile = (file: File) => ({
+  id: file.filePath,
+  type: 'file',
+  value: file,
+});
+
+export const nodeFromAsset = (asset: Asset) => ({
+  id: asset.hash,
+  type: 'asset',
+  value: asset,
+});
+
+const getFileNodesFromGraph = (graph: Graph): Array<File> => {
+  return Array.from(graph.nodes.values())
+    .filter((node: any) => node.type === 'file')
+    .map(node => node.value);
+}
+
+const getDepNodesFromGraph = (graph: Graph): Array<Dependency> => {
+  return Array.from(graph.nodes.values())
+    .filter((node: any) => node.type === 'dependency')
+    .map(node => node.value);
+}
+
+type DepUpdates = {
+  newFile?: File,
+  prunedFiles: Array<File>,
+}
+
+type FileUpdates = {
+  newDeps: Array<Dependency>,
+  prunedFiles: Array<File>
+}
+
+type AssetGraphOpts = {
+  entries: Array<string>,
+  rootDir: string,
+}
+
+/**
+ * AssetGraph is a Graph with some extra rules.
+ *  * Nodes can only have one of the following types "root", "dependency", "file", "asset"
+ *  * There is one root node that represents the root directory
+ *  * The root note has edges to dependency nodes for each entry file
+ *  * A dependency node should have an edge to exactly one file node
+ *  * A file node can have one to many edges to asset nodes which can have zero to many edges dependency nodes
+ */
+export default class AssetGraph extends Graph {
+  incompleteNodes: Map<NodeId, Node>;
+
+  constructor({ entries, rootDir }: AssetGraphOpts) {
+    super();
+    this.incompleteNodes = new Map();
+    this.initializeGraph({ entries, rootDir });
+  }
+
+  initializeGraph({ entries, rootDir }: AssetGraphOpts) {
+    let rootNode = nodeFromRootDir(rootDir);
+    this.addNode(rootNode);
+
+    let depNodes = entries.map(entry => nodeFromDep({
+      sourcePath: rootDir,
+      moduleSpecifier: entry,
+    }));
+
+    this.updateDownStreamNodes(rootNode, depNodes);
```

Comment:
Yeah, I struggled naming this method. Here is the comment I left explaining the function:

> // Update a node's downstream nodes making sure to prune any orphaned branches
> // Also keeps track of all added and removed edges and nodes

Basically, it's saying these are all of the nodes this node should be connected *to* (as opposed to from). It may already be connected to some of those nodes and it may be connected to nodes that it should no longer be connected to, so it needs to handle cleaning up edges and not duplicating any along with adding new ones.

Got any ideas on a better name? I'll try brainstorming too.